Add LXD and OpenStack image documentation (#289)
canonical/ubuntu-cloud-docs: public-images/public-images-explanation/lxd-openstack-images.rst

``<release>-<type>-cloudimg-<architecture>-<artefact>``.

LXD and OpenStack also have `Minimal Cloud
Images <https://cloud-images.ubuntu.com/minimal/>`__:
Ubuntu images that have a reduced runtime footprint, optimised kernel
and boot process. They are smaller and boot faster, but are not designed
for environments requiring human interaction or debugging.

LXD images
~~~~~~~~~~

To import an image into LXD, you will need two artefacts:

1. An LXD tarball:

   - The :ref:`lxd-tarball-ref` artefact has the extension ``*.lxd.tar.xz``
   - It contains the metadata needed by LXD to instantiate a container or virtual machine as well as a folder for any custom templates

2. A file system for a container or a bootable disk image for a virtual
   machine:

   - The file system for a container can be either a 

     * :ref:`root-tarball-ref` (``*-root.tar.xz``), or a 

     * :ref:`squashfs-ref` (``*.squashfs``)

   - The bootable disk image for a virtual machine is a :ref:`qcow-ref` image (``*.img``)

The following is an example command to import an LXD container image based on
downloaded Ubuntu 24.04 artefacts:

.. code :: bash

   lxc image import noble-server-cloudimg-amd64-lxd.tar.xz

>    noble-server-cloudimg-amd64-root.tar.xz --alias noble_container

Reviewer: Is it worth showing an example for lxd using a VM and .img file as well?

`OpenStack <https://ubuntu.com/openstack>`__ is an
open-source cloud platform designed to create and manage cloud
resources. By aggregating physical resources such as distributed
compute, network, and storage into a pool, OpenStack then allocates
virtual resources on-demand to users out of this pool. It does not
handle virtualisation itself, but acts as a wrapper that leverages
existing virtualisation technologies.

What are these images?
----------------------

Canonical provides cloud image artefacts on
`cloud-images.ubuntu.com <https://cloud-images.ubuntu.com/>`__ that have
been customised to run on public clouds, including LXD and OpenStack. To
learn more about these artefacts and supported architectures, visit our
Ubuntu cloud image artefacts documentation :doc:`here <../public-images-reference/artefacts>`.

How do you access them?
-----------------------

Go to `cloud-images.ubuntu.com <https://cloud-images.ubuntu.com/>`__ and
select a release. For the latest LTS release, Ubuntu 24.04 LTS Noble
Numbat, you would navigate to ``noble > current``. Note that all artefacts are
architecture specific, in the format
``<release>-<type>-cloudimg-<architecture>-<artefact>``.

> LXD and OpenStack also have `Minimal Cloud

Reviewer 2: To maintain consistency with the rest of the documentation set, can we remove the capitalisation and keep it as 'minimal cloud images'?

Images <https://cloud-images.ubuntu.com/minimal/>`__:
Ubuntu images that have a reduced runtime footprint, optimised kernel
and boot process. They are smaller and boot faster, but are not designed
for environments requiring human interaction or debugging.

LXD images
~~~~~~~~~~

To import an image into LXD, you will need two artefacts:

> 1. An LXD tarball:

Reviewer 2: ```suggestion
1. A LXD tarball:
```
since it is pronounced lex-dee :) (https://documentation.ubuntu.com/server/how-to/containers/lxd-containers/)

   - The :ref:`lxd-tarball-ref` artefact has the extension ``*.lxd.tar.xz``
   - It contains the metadata needed by LXD to instantiate a container or virtual machine as well as a folder for any custom templates

2. A file system for a container or a bootable disk image for a virtual
   machine:

   - The file system for a container can be either a 

     * :ref:`root-tarball-ref` (``*-root.tar.xz``), or a 

     * :ref:`squashfs-ref` (``*.squashfs``)

   - The bootable disk image for a virtual machine is a :ref:`qcow-ref` image (``*.img``)

> The following are example commands to import LXD containers and virtual machines, respectively, based on downloaded Ubuntu 24.04 artefacts:

Reviewer 2: ```suggestion
The following are example commands to import an image into LXD containers and virtual machines, respectively, based on downloaded Ubuntu 24.04 artefacts:
```
Does that make sense? Or maybe something like 'to import an image to create LXD containers and virtual machines'?

Author: Good catch here, I think some context got lost when I was reworking the sentence. This makes sense to me: "The following are example commands to import _an image for creating_ LXD containers and virtual machines based on downloaded Ubuntu 24.04 artifacts."

Reviewer 2: You could have directly included that! :)

Author: I thought I had :) Wrote both down in the editor to compare and must have deleted the new one instead of the old.

.. code :: bash

   lxc image import noble-server-cloudimg-amd64-lxd.tar.xz \
      noble-server-cloudimg-amd64-root.tar.xz --alias noble_container

.. code :: bash

   lxc image import noble-server-cloudimg-amd64-lxd.tar.xz \
      noble-server-cloudimg-amd64.img --alias noble_vm

OpenStack images
~~~~~~~~~~~~~~~~

OpenStack uses QCOW images. Download the artefact for your chosen
architecture with the ``*.img`` extension.

Use the `OpenStack command-line
client <https://docs.openstack.org/ocata/admin-guide/common/cli-install-openstack-command-line-clients.html>`__
to interact with OpenStack. An example of uploading an image looks 
something like this:

.. code :: bash

   openstack image create “Ubuntu-24.04” \
      --file noble-server-cloudimg-amd64.img \
      --disk-format qcow2 \
      --container-format bare \
      --public

> To learn more about managing images with OpenStack, you can read their

Reviewer 2: ```suggestion
To learn more about managing images with OpenStack, you can refer to the `Manage images`_ section of their
```
We try to avoid using 'here' for link descriptions and instead use the expected content title as is. Since the whole page seems relevant I'm using the page title (& a link to it). If you think the specific section that you pointed to makes more sense, we can use that too.

Author: Good to know, `here` felt awkward.

documentation

> `here <https://docs.openstack.org/ocata/admin-guide/common/cli-manage-images.html#create-or-update-an-image-glance>`__.

Author: I've added the `.. LINKS` section to the bottom of the page and will move all external links there.

> Configuration

Reviewer 2: ```suggestion
How do you configure them?
```

> -------------

Reviewer 2: Increase this length suitably.

Configuring an Ubuntu cloud image allows you to make changes that tailor
the image to your specific use case. You can automate the creation of
user accounts, configure SSH access, or install software before the
instance starts.

LXD images
~~~~~~~~~~

You can configure your cloud images in LXD either before you import them
or after. Configuring your images before importing them is most commonly
done by editing the ``metadata.yaml`` file contained in the LXD tarball.
Configuring your images after importing them is done through the CLI.

If you are interested in configuration of LXD containers rather than
images, take a look at the Ubuntu Server documentation

> `here <https://documentation.ubuntu.com/server/how-to/containers/lxd-containers/>`__.

Reviewer 2: Could you do the same thing with the link here? (Change 'here' and move the link to the end of the file)

> Metadata

Reviewer 2: ```suggestion
Configuring metadata
```

^^^^^^^^

LXD metadata is stored in the ``metadata.yaml`` file in the LXD tarball. This file contains all of 
the information needed to run an image in LXD. To make changes to this file, you will have to:

1. Uncompress the LXD tarball.
2. Make modifications to the ``metadata.yaml`` file. See the `LXD image
   format
   documentation <https://documentation.ubuntu.com/lxd/en/latest/reference/image_format/#>`__
   to learn more about image metadata and the templates you may wish to
   modify.
3. Compress the metadata and templates.

This snippet from the `How to customise LXD image metadata for

> cloud-init <https://discourse.ubuntu.com/t/how-to-customize-lxd-image-metadata-for-cloud-init/25157>`__

Reviewer 2: Could you move this link also to the end, so that it can be reused in the next section? 
(In fact, if you could move all links to the end, it would be great!)

guide referenced below demonstrates a typical workflow:

.. code :: bash

   # Uncompress original LXD metadata
   $ tar xf ${RELEASE}-server-cloudimg-amd64-lxd.tar.xz
   # Add directives to create /etc/cloud/cloud.cfg.d/95-use-lxd.cfg
   $ cat > templates/cloud-init-use-lxd.tpl <<EOF
   # Added by LXD metadata.yaml
   datasource_list: [ LXD, NoCloud ]
   EOF
   $ cat > add-lxd.yaml <<EOF
       /etc/cloud/cloud.cfg.d/95-use-lxd.cfg:
           when:
               - create
               - copy
           template: cloud-init-use-lxd.tpl
   EOF
   $ cat add-lxd.yaml >> metadata.yaml
   # Compress LXD metadata and templates
   $ tar -czf ${RELEASE}-server-cloudimg-amd64-prefer-lxd.tar.xz metadata.yaml templates/

> Cloud-init

Reviewer 2: ```suggestion
Configuring cloud-init
```

^^^^^^^^^^

`Cloud-init <https://cloudinit.readthedocs.io/en/latest/index.html>`__
is used to initialise cloud instances on first boot. A detailed guide on
customising LXD image metadata for ``cloud-init`` can be found

> `here <https://discourse.ubuntu.com/t/how-to-customize-lxd-image-metadata-for-cloud-init/25157>`__.

Reviewer 2: Change the 'here' and reuse the link from the previous section.

If you want to configure ``cloud-init`` once an instance has been
created (but not booted), refer to the LXD docs

> `here <https://documentation.ubuntu.com/lxd/en/latest/cloud-init/>`__.

Reviewer 2: Change the 'here'.

> CLI

Reviewer 2: ```suggestion
Configuring after import using CLI
```

^^^

The `LXD documentation on
images <https://documentation.ubuntu.com/lxd/en/latest/howto/images_manage/>`__
has an extensive guide on managing images. Two of the most common use
cases are to set individual properties or to edit all of the image
properties.

.. code :: bash

   # set a specific image property
   lxc image set-property <image_ID> <key> <value>

.. code :: bash

   # edit the full image properties
   lxc image edit <image_ID>

OpenStack images
~~~~~~~~~~~~~~~~

OpenStack has an extensive guide on `modifying
images <https://docs.openstack.org/image-guide/modify-images.html>`__
that is applicable to the QCOW images Ubuntu provides. It leverages ``libguestfs`` 
`tools <https://libguestfs.org/>`__ in order to access and
modify disk images. You can use the ``guestfish``
`interactive shell <https://libguestfs.org/guestfish.1.html>`__ 

> — which exposes the full functionality of the ``guestfs`` API — or

Reviewer 2: ```suggestion
(which exposes the full functionality of the ``guestfs`` API) or
```

> rely on the ``virt-*`` tools from ``libguestfs`` — like ``virt-cat`` for

Reviewer 2: ```suggestion
rely on the ``virt-*`` tools from ``libguestfs`` to perform specific tasks. For instance, use ``virt-cat`` for
```

displaying files, ``virt-df`` for checking free space, and

> ``virt-inspector`` for inspecting VM images — to perform specific tasks.

Reviewer 2: ```suggestion
``virt-inspector`` for inspecting VM images.
```

Author: Whoops, many changes to this sentence and this one slipped.

I usually fix in one commit and then rebase to keep the commit history clean, and haven't looked into doing that via the GitHub UI.

Reviewer 2: Using the `Add suggestion to batch` in the UI option allows you to create a single commit. But yeah you'll still need to squash commits to keep the commit history clean and run the spell/link checks locally.
